src/appleseed/renderer/modeling/camera/fisheyelenscamera.cpp

```diff
@@ -0,0 +1,511 @@
+                get_path().c_str(),
+                Model,
+                m_film_dimensions[0],
+                m_film_dimensions[1],
+                m_focal_length,
+                m_near_z,
+                m_shutter_open_begin_time,
+                m_shutter_open_end_time,
+                m_shutter_close_begin_time,
+                m_shutter_close_end_time,
+                projection_type);
+        }
+
+        bool on_render_begin(
+            const Project&      project,
+            IAbortSwitch*       abort_switch) override
+        {
+            if (!Camera::on_render_begin(project, abort_switch))
+                return false;
+
+            // Extract the film dimensions from the camera parameters.
+            m_film_dimensions = extract_film_dimensions();
+
+            // Extract the focal length from the camera parameters.
+            m_focal_length = extract_focal_length(m_film_dimensions[0]);
+
+            // Extract the abscissa of the near plane from the camera parameters.
+            m_near_z = extract_near_z();
+
+            // Precompute reciprocals of film dimensions.
+            m_rcp_film_width = 1.0 / m_film_dimensions[0];
+            m_rcp_film_height = 1.0 / m_film_dimensions[1];
+
+            // Precompute pixel area.
+            const size_t pixel_count = project.get_frame()->image().properties().m_pixel_count;
+            m_pixel_area = m_film_dimensions[0] * m_film_dimensions[1] / pixel_count;
+
+            const string projection_type = m_params.get_required<string>("projection_type", "equisolid_angle");
+
+            if (projection_type == "equisolid_angle")
+                m_projection_type = EquisolidAngle;
+
+            else if (projection_type == "equidistant")
+                m_projection_type = Equidistant;
+                
+            else if (projection_type == "stereographic")
+                m_projection_type = Stereographic;
+
+            else if (projection_type == "thoby")
+                m_projection_type = Thoby;
+
+            else
+            {
+                RENDERER_LOG_ERROR(
+                    "invalid value \"%s\" for parameter \"projection_type\", "
+                    "using default value \"equisolid_angle\".",
+                    projection_type.c_str());
+                m_projection_type = EquisolidAngle;
+            }
+
+            return true;
+        }
+
+        //
+        // Transform ray direction in selected projection.
+        // Reference : 
+        //   https://wiki.panotools.org/Fisheye_Projection
+        //   https://de.wikipedia.org/wiki/Fischaugenobjektiv
+        //   http://michel.thoby.free.fr/Fisheye_history_short/Projections/Fisheye_projection-models.html
+        //
+
+        void transform_ray_dir(Vector3d& dir) const
+        {
+            // Transform the direction vector to be in the xz plane.
+            const Transformd rot = Transformd(Matrix4d::make_rotation_z(-atan2 (dir[1], dir[0])));
```

Comment:
Remove space between `atan2` and opening parenthesis.

I also think this code should be unrolled, simplified and optimized. Constructing a matrix, then a transform, then transforming is a lot of redundant operations.

Especially since this constructor of `foundation::Transform` is actually inverting a matrix...

Reply:
Should it be like this? 

https://github.com/appleseedhq/appleseed/blob/adbcd712658c818c7df703ddeece0e997e4c9e00/src/appleseed/foundation/meta/tests/test_transform.cpp#L162

or is it problem of my mathematical approach?

Reply:
I only had a quick look but mathematically your approach seems correct. The problem is that there are a lot of hidden operations behind that single line of code, including a full matrix inversion. You don't need all that.

Reply:
You should look at how `renderer::Transform` works and extracts just what you need from it. You can also probably simplify things a bit since you only have a rotation around the Z axis.

Reply:
I changed my approach (without using transform), and it seems good to me.